experimental/hip/dynamic_symbols.c

```diff
@@ -49,24 +50,112 @@ static iree_status_t iree_hal_hip_dynamic_symbols_resolve_all(
   return iree_ok_status();
 }
 
+static bool iree_hal_hip_try_load_dylib(const char* file_path,
+                                        iree_dynamic_library_flags_t flags,
+                                        iree_allocator_t allocator,
+                                        iree_status_t* pending_fail_status,
+                                        iree_dynamic_library_t** out_library) {
+  iree_status_t status = iree_dynamic_library_load_from_file(
+      file_path, flags, allocator, out_library);
+  if (iree_status_is_ok(status)) {
+    return true;
+  }
+
+  char* buffer = NULL;
+  iree_host_size_t length = 0;
+  if (iree_status_to_string(status, &allocator, &buffer, &length)) {
+    *pending_fail_status = iree_status_annotate_f(
+        *pending_fail_status, "\n  Tried: %s\n    %s", file_path, buffer);
+    iree_allocator_free(allocator, buffer);
+  }
+
+  iree_status_ignore(status);
+  return false;
+}
+
 iree_status_t iree_hal_hip_dynamic_symbols_initialize(
-    iree_allocator_t host_allocator, iree_hal_hip_dynamic_symbols_t* out_syms) {
+    iree_allocator_t host_allocator, iree_host_size_t hip_lib_search_path_count,
+    const iree_string_view_t* hip_lib_search_paths,
+    iree_hal_hip_dynamic_symbols_t* out_syms) {
   IREE_ASSERT_ARGUMENT(out_syms);
   IREE_TRACE_ZONE_BEGIN(z0);
 
   memset(out_syms, 0, sizeof(*out_syms));
-  iree_status_t status = iree_dynamic_library_load_from_files(
-      IREE_ARRAYSIZE(iree_hal_hip_dylib_names), iree_hal_hip_dylib_names,
-      IREE_DYNAMIC_LIBRARY_FLAG_NONE, host_allocator, &out_syms->dylib);
-  if (iree_status_is_not_found(status)) {
-    iree_status_ignore(status);
-    status = iree_make_status(
-        IREE_STATUS_UNAVAILABLE,
-        "HIP runtime library 'amdhip64.dll'/'libamdhip64.so' not available;"
-        "please ensure installed and in dynamic library search path");
+
+  // Load the library.
+  bool loaded_one = false;
+  iree_status_t status = iree_ok_status();
+  iree_status_t pending_fail_status = iree_make_status(
```

Comment:
status as flow control isn't great as it captures backtraces, may be used as a break-on-first-error, etc - instead, use an iree_string_builder_t to generate your message and then only create the status with it at the end

it also makes lifetime management clearer (it's easy to lose track of statuses) and iree_string_builder_t is much more efficient (zero allocations of nothing appended, grows in chunks instead of allocating one per status annotation, etc)

Reply:
Switched to string builder.